Convex_hull_3/include/CGAL/Convex_hull_3/predicates.h

```diff
@@ -14,267 +14,154 @@
       default : {
-                  Vec3<NT> avg;
-                  for( const SphericalPolygonElement<NT> & v : *this )
+                  Vector_3 avg(0.0f, 0.0f, 0.0f);
+                  for( const SphericalPolygonElement & v : *this )
                     avg = avg + v.vertex_;
                   return avg;
                 } break;
     }
   }
 
-  void clip(const Vec3<NT> & OrigVertex, SphericalPolygon<NT> & result, bool doClean=true) const {
+  void clip(const Vector_3 & OrigVertex, SphericalPolygon & result, bool doClean=true) const {
     // PRECONDITION : clipNorth, and all northes are normalized.
 #define _ray_spherical_eps 1e-6f
+    typename K::Collinear_3 compute_collinear =  K().collinear_3_object();
+    typename K::Angle_3 compute_angle = K().angle_3_object();
     const int n = this->size();
     result.clear();
     switch( n ) {
       case 0 : break; // 0 means empty, so nothing to do
       case 1 : {
                  result = (*this);
-                 Vec3<NT> clipNorth = OrigVertex.normalized();
-                 NT dot = this->begin()->north_ | clipNorth;
+                 Vector_3 clipNorth = OrigVertex / std::sqrt(OrigVertex.squared_length());
+#ifdef USE_EXACT_PREDICATE
+                 bool collinear = compute_collinear(ORIGIN + this->begin()->north_, ORIGIN, ORIGIN + OrigVertex);
+                 if (collinear) {
+                    Angle angle = compute_angle(this->begin()->north_, OrigVertex);
+                    if ( angle == OBTUSE ) {
+                      result.clear();
+                      break;
+                    } if ( angle == ACUTE ) {
+                      break;
+                    }
+                 }
+#else
+                 auto dot = this->begin()->north_ * clipNorth;
                  if( dot < -0.99984769515 ) { // about one degree
                    // intersection of two almost opposite hemispheres ==> empty
                    result.clear();
                    break;
                  } else if( dot > 0.99984769515 ) {
                    break;
                  }
-                 Vec3<NT> v(Vec3<NT>::cross(clipNorth, this->begin()->north_).LInfNormalized());
+#endif
+                 Vector_3 v = LInfNormalize(cross_product(OrigVertex, this->begin()->north_));
                  result.begin()->vertex_ = v;
-                 result.emplace_back(v.negated(), clipNorth);
+                 result.emplace_back(-v, clipNorth);
                  break;
                }
       case 2 : {
                  result = (*this);
-                 Vec3<NT> clipNorth = OrigVertex.LInfNormalized();
+                 Vector_3 clipNorth = LInfNormalize(OrigVertex);
                  iterator next = result.begin();
                  iterator cur = next++;
-                 NT vDot = this->begin()->vertex_ | clipNorth;
-                 if( vDot >= _ray_spherical_eps ) {
+#ifdef USE_EXACT_PREDICATE
+                 Angle angle = compute_angle(this->begin()->vertex_, OrigVertex);
```

Comment:
You can directly use https://doc.cgal.org/latest/Kernel_23/classKernel_1_1Angle__3.html, then you avoid a construction